src/client/Microsoft.Identity.Client/Extensibility/CdtAuthenticationScheme.cs

```diff
@@ -0,0 +1,296 @@
+            var body = new JObject
+            {
+                // Mandatory parameters
+                [CdtClaimTypes.Ticket] = secret,
+                [CdtClaimTypes.ConstraintsToken] = CreateCdtConstraintsJwT(nonce)
+                //[CdtClaimTypes.ConstraintsToken] = string.IsNullOrEmpty(encryptionKey) 
+                //                                    ? CreateCdtConstraintsJwT(msalAccessTokenCacheItem) :
+                //                                      CreateEncryptedCdtConstraintsJwT(msalAccessTokenCacheItem, encryptionKey)
+            };
+
+            return body;
+        }
+
+        //private JToken CreateEncryptedCdtConstraintsJwT(MsalAccessTokenCacheItem msalAccessTokenCacheItem, string encryptionKey)
+        //{
+        //    var header = new JObject();
+        //    header[JsonWebTokenConstants.Algorithm] = Constants.CdtEncryptedAlgoryth;
+        //    header[JsonWebTokenConstants.CdtEncrypt] = Constants.CdtEncryptedValue;
+
+        //    var body = new JObject
+        //    {
+        //        // TODO: ENCRYPT JWT
+        //        [CdtClaimTypes.Constraints] = CreateCdtConstraintsJwT(msalAccessTokenCacheItem)
+        //    };
+
+        //    string cdtConstraintToken = CreateJWS(JsonHelper.JsonObjectToString(body), JsonHelper.JsonObjectToString(header));
+        //    return cdtConstraintToken;
+        //}
+
+        private JToken CreateCdtConstraintsJwT(string nonce)
+        {
+            var header = new JObject();
+            header[JsonWebTokenConstants.Algorithm] = _cdtCryptoProvider.CryptographicAlgorithm;
+            header[JsonWebTokenConstants.Type] = JasonWebTokenType;
+
+            var body = new JObject
+            {
+                // Mandatory parameters
+                [CdtClaimTypes.Nonce] = nonce,
+                [CdtClaimTypes.Constraints] = _constraints
+            };
+
+            string cdtConstraintToken = CreateJWS(JsonHelper.JsonObjectToString(body), JsonHelper.JsonObjectToString(header));
+            return cdtConstraintToken;
+        }
+
+        /// <summary>
+        /// A key ID that uniquely describes a public / private key pair. While KeyID is not normally
+        /// strict, AAD support for Cdt requires that we use the base64 encoded JWK thumbprint, as described by 
+        /// https://tools.ietf.org/html/rfc7638
+        /// </summary>
+        private static byte[] ComputeThumbprint(string canonicalJwk)
+        {
+            using (SHA256 hash = SHA256.Create())
+            {
+                return hash.ComputeHash(Encoding.UTF8.GetBytes(canonicalJwk));
+            }
+        }
+
+        /// <summary>
+        /// Creates a JWS (json web signature) as per: https://tools.ietf.org/html/rfc7515
+        /// Format: header.payload.signed_payload
+        /// </summary>
+        private string CreateJWS(string payload, string header, bool signPayload = true)
+        {
+            StringBuilder sb = new StringBuilder();
```

Comment:
can use spans here to improve performance.

Reply:
This should be done with Wilson (as shown above)